feat: Report tag

YOGESH-08 opened the pull request:
### Added a complete tag-reporting system

- Implemented a backend API with strict validation and IP-based rate limiting.
- Created a new MongoDB model (TagReport) to store user-submitted corrections for subject, exam, slot, year, and course code.
- Built a frontend modal to allow users to report incorrect tags directly from the PDF viewer.

### Updated PDF Viewer UI

- Removed page up and down buttons.
- Added a new red report flag button in both mobile and desktop toolbars to quickly report tagging errors.

## 📌 Purpose
- This PR introduces a full reporting workflow so users can flag wrong metadata (subject, exam, slot, year) for any paper.
It improves data accuracy and provides a structured way to collect user feedback on incorrect tags. 
- Also Prevents spamming of report by a single user on a single paper, by IP rate limit, and Maximum Report Rate limit for Papers.

---

## Corresponding issue: closes #284 

---

## 🖼️ Showcase

<img width="1188" height="995" alt="Screenshot 2025-11-26 144656" src="https://github.com/user-attachments/assets/d29452d4-db1f-41e3-8503-612917160abe" />
<img width="1613" height="870" alt="Screenshot 2025-11-26 144611" src="https://github.com/user-attachments/assets/f16c3918-5cb5-4d10-9b9e-99d247b10ae4" />

<img width="1336" height="662" alt="Screenshot 2025-11-26 145724" src="https://github.com/user-attachments/assets/a1c3b345-0bef-4578-a589-65c67a4e92a3" />

---

## 🔧 Changes

- Added /api/report-tag endpoint with:
    - Allowed fields validation
    - Exam/year format validation
    - IP + paper-based rate limiting
    - Max reports per paper
- Added new Mongoose model TagReport.
- Added ReportButton component to trigger the modal.
- Added ReportTagModal with:
    - Multi-select for incorrect fields
    - Smart input loading from original values
    - Slot/year validation
    - Only changed fields sent to backend
- Cleaned up PDF viewer UI: removed page navigation arrows, added report flag.

- No. of reports from an IP, in an hour is set to three and Maximum reports allowed for a paperID in DB is set to two.( Only For Testing).

---

## ➕ Additional Notes

- Current report limit values are set low for testing and can be increased in production. No. of reports from an IP, in an hour is set to be three and Maximum reports allowed for a paperID is set to be two.( Only For Testing)
- Next steps : build an admin dashboard to review and resolve reports.
- No breaking changes — viewer functionality remains intact aside from toolbar UI cleanup.

YOGESH-08 committed:
feat: Report tag

added a complete tag-reporting system->a backend API with validation + rate limiting, a new DB model (tagreports) to store reports, and a frontend modal.

Removed page up and down buttons , replaced by report Flag.

abhitrueprogrammer commented on src/components/ReportTagModal.tsx:
```diff
@@ -0,0 +1,411 @@
+    <Dialog open={modalOpen} onOpenChange={modalSetOpen}>
+      {!isControlled && (
+        <DialogTrigger asChild>
+          <button className="rounded-full border border-red-300 p-2 text-red-500 transition hover:border-red-500 hover:text-red-600 hover:shadow-[0_0_8px_rgba(255,0,0,0.4)]">
+            <FaFlag className="text-sm" />
+          </button>
```
use shadCN components for consistency. EG here use Button instead of button

abhitrueprogrammer commented on src/components/ReportTagModal.tsx:
```diff
@@ -0,0 +1,411 @@
+  if (reportedFields.length === 0) {
+    toast.error("You haven’t changed anything to report.");
+    return;
+  }
```
just disable the button if user haven't changed anything right?

YOGESH-08 replied:
yeah sure

abhitrueprogrammer commented on src/components/ReportTagModal.tsx:
```diff
@@ -0,0 +1,411 @@
+              onChange={(e) => setEmail(e.target.value)}
+              placeholder="you@example.com"
+            />
+          </div>
```
this pattern of taking in Input for and label is being repeated a lot right? Could be that you separate this into reusable function or component?

YOGESH-08 replied:
sure

abhitrueprogrammer commented on src/components/pdfViewer.tsx:
```diff
@@ -302,8 +293,8 @@ export default function PdfViewer({ url, name }: PdfViewerProps) {
-                    renderAnnotationLayer
-                    renderTextLayer
+                    renderAnnotationLayer={false}
+                    renderTextLayer={false}
```
why do this 👀

YOGESH-08 replied:
Actually I was facing an error that,

""Cannot find module or type declarations for side-effect import of 'react-pdf/dist/Page/TextLayer.css' annot find module or type declarations for side-effect import of 'react-pdf/dist/Page/AnnotationLayer.css'.""

and tried to resolve this, but it was crashing, so i found another solution and that was this, and everything was fine when i am using it.

I will look into that once, bring back the old one

YOGESH-08 commented on src/app/api/report-tag/route.ts:
```diff
@@ -0,0 +1,140 @@
+    return url.hostname || "127.0.0.1";
+  } catch {
+    return "127.0.0.1";
+  }
```
instead of using "x-forwarded-for" and "x-real-ip"(which can be spoofed) -->  req.ip is being used to get the actual ip 
### Additionally 
added a per-paper report limit, so even if someone tries to bypass the IP limit, each paper can only receive a small fixed number of reports.

YOGESH-08 committed:
fix: Upstash Rate Limiter

1) Fixed a few minor issues.
2) Used existing Shadcn components to       maintain UI consistency.
3) Implemented Upstash Rate Limiting and set up a Redis database to prevent spoofing(IP) and report spamming.

Abh1noob commented on src/app/api/report-tag/route.ts:
```diff
@@ -0,0 +1,95 @@
+    }
+    const reportedFields = (body.reportedFields ?? [])
+      .map((r: any) => ({
+        field: String(r.field).trim(),
```
typecast it properly. avoid using `any`

Abh1noob commented on src/app/api/report-tag/route.ts:
```diff
@@ -0,0 +1,95 @@
+import { Ratelimit } from "@upstash/ratelimit";
+import { redis } from "@/lib/utils/redis";
+
+const ALLOWED_EXAMS = ["CAT-1", "CAT-2", "FAT"];
```
we must already have allowed exams defined elsewhere. re-use that definition to reduce duplication and improve consistency

YOGESH-08 replied:
Sure

Abh1noob commented on src/app/api/report-tag/route.ts:
```diff
@@ -0,0 +1,95 @@
+  analytics: true,
+});
+
+function getClientIp(req: any): string {
```
use appropriate type

Abh1noob reviewed (changes requested):
good job for first PR

i suggest you to use contexts instead of propdrilling paper information from pdfviewer till the very child element

(its just about preference)

YOGESH-08 committed:
fix: in reportTag files

1) usecontext instead of propdrilling paper information from pdfviewer till the very child element

2) separate reusable component for Label and input

3)disabled the button if user didn't change anything.instead of throwing an error.

abhitrueprogrammer commented:
good job, implement the few changes I mentioned on whatsapp nd make a new PR for those
